### PR TITLE
Adding a control flow to remove approval label from PRs when reviews are dismissed.

### PR DESCRIPTION
**Pull request by ben-ellevation**
As it stands right now, PRs retain the approval label even when approvals are dismissed. This'll fix that and make our deploy measures more accurate.

### CODE REVIEW

File: entrypoint.sh [hunk 1]
```diff
@@ -76,8 +76,51 @@ label_when_approved() {
...
+      fi
+
+      break
+    fi
+  done
+}
+
 if [[ "$action" == "submitted" ]] && [[ "$state" == "approved" ]]; then
   label_when_approved
+elif [[ "$action" == "dismissed" ]]; then
```
Comment (ben-ellevation) [inline entrypoint.sh:122]: I _think_ this is the correct combination of sorcerous action/state, but I think the worst case scenario is it doesn't fire.

Reference:
https://docs.github.com/en/developers/webhooks-and-events/webhooks/webhook-events-and-payloads#pull_request_review

File: entrypoint.sh [hunk 2]
```diff
@@ -76,8 +76,51 @@ label_when_approved() {
...
+            -X DELETE \
+            "${URI}/repos/${GITHUB_REPOSITORY}/issues/${number}/labels/${addLabel}"
+      fi
+
+      break
+    fi
+  done
+}
+
 if [[ "$action" == "submitted" ]] && [[ "$state" == "approved" ]]; then
```
Comment (ben-ellevation) [inline entrypoint.sh:120]: I'm a little annoyed that I couldn't find anything for state in the docs. :|

**Review by justin-ellevation — APPROVED** [review]

File: entrypoint.sh [hunk 3]
```diff
@@ -76,8 +76,51 @@ label_when_approved() {
   done
 }
 
+label_when_disapproved() {
```
Comment (ben-ellevation) [inline entrypoint.sh:79]: Most of this is copy/pasta, except for the `lt` if statement and swapping out the labels to add/remove.

File: entrypoint.sh [hunk 4]
```diff
@@ -76,8 +76,51 @@ label_when_approved() {
...
+
+      curl -sSL \
+        -H "${AUTH_HEADER}" \
+        -H "${API_HEADER}" \
+        -X POST \
+        -H "Content-Type: application/json" \
+        -d "{\"labels\":[\"${REMOVE_LABEL}\"]}" \
+        "${URI}/repos/${GITHUB_REPOSITORY}/issues/${number}/labels"
+
+      if [[ -n "$addLabel" ]]; then
```
Comment (andy-ellevation) [inline entrypoint.sh:107]: I don't think you need this if statement. in the other code, this was checking to see if $REMOVE_LABEL was defined before deleting the label. i think you should put this one outside the block, and then use the same -n check on $REMOVE_LABEL before trying to add it. alternately, you could require $REMOVE_LABEL in the initiation above, using a check like at line 20.

File: entrypoint.sh (hunk 2 above)
Comment (andy-ellevation) [reply entrypoint.sh:120]: there's some here in the 'event' property: https://docs.github.com/en/rest/reference/pulls#create-a-review-for-a-pull-request
and if you search in the example responses, you can see state examples.

File: entrypoint.sh (hunk 1 above)
Comment (andy-ellevation) [reply entrypoint.sh:122]: You could consider running the add and delete within the same event: if there are at least $APPROVALS approved reviews, then add the label. if there are less, then remove it. you'd still want to fire on the dismissed action.

you could also consider firing on https://docs.github.com/en/developers/webhooks-and-events/webhooks/webhook-events-and-payloads#pull_request
and use at least these actions:
ready_for_review
reopened
review_request_removed
review_requested

**Review by andy-ellevation — CHANGES_REQUESTED** [review]
your swapping of the labels makes the $REMOVE_LABEL a required parameter, where it wasn't before.

File: entrypoint.sh (hunk 1 above)
Comment (ben-ellevation) [reply entrypoint.sh:122]: 5 PM Ben was originally going for least ~amount of effort~ chances of disrupting the current behavior, but yeah, I'll DRY this up a bit. No reason I can't just call the same function.

**Comment by ben-ellevation** [comment]
Oh yeesh wait a sec I think I confused myself on the code. One sec. (I thought I removed a function but apparently I just didn't?)

**Comment by ben-ellevation** [comment]
Okay all good!

File: entrypoint.sh [hunk 5]
```diff
@@ -52,31 +59,37 @@ label_when_approved() {
 
     echo "${approvals}/${APPROVALS} approvals"
 
+    label_to_add = addLabel
+    label_to_remove = $REMOVE_LABEL
```
Comment (ben-ellevation) [inline entrypoint.sh:63]: I figured shell has ACTUAL SCOPING unlike Python, but if it's Python scoping then I can move those into the if block below. Could also set these to a NULL if preferred?

File: entrypoint.sh [hunk 6]
```diff
@@ -52,31 +59,37 @@ label_when_approved() {
...
+      -H "${API_HEADER}" \
+      -X DELETE \
+      "${URI}/repos/${GITHUB_REPOSITORY}/issues/${number}/labels/${label_to_remove}"
   done
 }
 
 if [[ "$action" == "submitted" ]] && [[ "$state" == "approved" ]]; then
+  # APPROVE Branch
+  label_when_approved
+elif ( [[ "$action" == "submitted" ]] || [[ "$action" == "dismissed" ]] ) && ( [[ "$state" == "DISMISSED" ]] || [[ "$state" == "CHANGES_REQUESTED" ]] ); then
```
Comment (ben-ellevation) [inline entrypoint.sh:91]: I opted to leave these branches separate for readability but I don't feel too strongly about this.

File: entrypoint.sh [hunk 7]
```diff
@@ -16,6 +16,7 @@ if [[ -z "$GITHUB_EVENT_PATH" ]]; then
   exit 1
 fi
 
+# ADD_LABEL: The label that will be add when the PR is approved (and the label that will be removed when the PR lacks approvals)
```
Comment (ben-ellevation) [inline entrypoint.sh:19]: I could rename these for clarity but I'd rather not change the interface right now.

File: entrypoint.sh [hunk 8]
```diff
@@ -52,31 +59,37 @@ label_when_approved() {
...
+      label_to_remove = addLabel
     fi
+
+    curl -sSL \
+      -H "${AUTH_HEADER}" \
+      -H "${API_HEADER}" \
+      -X POST \
+      -H "Content-Type: application/json" \
+      -d "{\"labels\":[\"${label_to_add}\"]}" \
+      "${URI}/repos/${GITHUB_REPOSITORY}/issues/${number}/labels"
```
Comment (ben-ellevation) [inline entrypoint.sh:78]: Also assuming adding a label that already exists or removing a label that doesn't exist doesn't cause a catastrophic failure on GitHub. >.>

Comment (andy-ellevation) [reply entrypoint.sh:78]: we'll find out!

File: entrypoint.sh (hunk 5 above)
Comment (andy-ellevation) [reply entrypoint.sh:63]: i think it does not have actual scoping, last i checked. also, there's a convention that variables are ALL CAPS.

File: entrypoint.sh (hunk 6 above)
Comment (andy-ellevation) [reply entrypoint.sh:91]: i like them being separate here. parsing actions is potentially annoying.

**Review by andy-ellevation — APPROVED** [review]

**Review by alla-ellevation — APPROVED** [review]

**Pull request merged**